linkerd/stack/src/gate.rs

```diff
@@ -1,81 +1,97 @@
 #[derive(Clone, Debug)]
-pub struct Tx(Arc<Shared>);
+pub struct Tx(Arc<watch::Sender<State>>);
 
-#[derive(Debug)]
-struct Shared {
-    open: AtomicBool,
-    notify: Notify,
-    closed: Notify,
+#[derive(Clone, Debug)]
+pub enum State {
+    /// The gate is open and unlimited.
+    Open,
+
+    /// The gate is limited by the provided semaphore. Permits are forgotten as
+    /// requests are processed so that this semaphore can specify a limit on the
+    /// number of requests to be admitted by the [`Gate`].
+    Limited(Arc<Semaphore>),
+
+    /// The gate is shut and no requests are to be admitted.
+    Shut,
 }
 
 /// Creates a new gate channel.
 pub fn channel() -> (Tx, Rx) {
-    let shared = Arc::new(Shared {
-        open: AtomicBool::new(true),
-        notify: Notify::new(),
-        closed: Notify::new(),
-    });
-    (Tx(shared.clone()), Rx(shared, Arc::new(())))
+    let (tx, rx) = watch::channel(State::Open);
+    (Tx(Arc::new(tx)), Rx(rx))
 }
 
+struct Permit(Option<OwnedSemaphorePermit>);
+
 // === impl Rx ===
 
 impl Rx {
     /// Indicates whether the gate is open.
+    pub fn state(&self) -> State {
+        self.0.borrow().clone()
+    }
+
     pub fn is_open(&self) -> bool {
-        self.0.open.load(std::sync::atomic::Ordering::Acquire)
+        matches!(self.state(), State::Open)
+    }
+
+    pub fn is_limited(&self) -> bool {
+        matches!(self.state(), State::Limited(_))
     }
 
-    /// Indicates whether the gate is closed.
-    #[inline]
     pub fn is_shut(&self) -> bool {
-        !self.is_open()
+        matches!(self.state(), State::Shut)
     }
 
     /// Waits for the gate state to change.
-    pub async fn changed(&self) -> bool {
-        self.0.notify.notified().await;
-        self.is_open()
+    pub async fn changed(&mut self) -> Result<(), watch::error::RecvError> {
+        self.0.changed().await
     }
 
     /// Waits for the gate state to change to be open.
-    pub async fn opened(&self) {
-        if self.is_open() {
-            return;
-        }
-
-        while !self.changed().await {}
-    }
-}
+    async fn acquire(&mut self) -> Option<OwnedSemaphorePermit> {
+        loop {
+            let state = self.0.borrow_and_update().clone();
+            match state {
+                State::Open => return None,
+                State::Limited(sem) => match sem.acquire_owned().await {
+                    Ok(permit) => return Some(permit),
+                    Err(_) => {
+                        debug!("Gate closed");
+                        return None;
+                    }
+                },
+                State::Shut => {}
+            }
 
-impl Drop for Rx {
-    fn drop(&mut self) {
-        let Rx(_, handle) = self;
-        if Arc::strong_count(handle) == 1 {
-            self.0.closed.notify_waiters();
+            if let Err(e) = self.0.changed().await {
+                debug!("Gate closed: {:?}", e);
+                futures::future::pending::<()>().await;
+            }
```

Comment:
if the `Tx` side of the gate is dropped, is it actually correct for this future to be pending forever? This will result in the `Gate` never opening, but it won't error out...should we have `poll_ready` return an error in this case?

Reply:
No, I don't think so.